sdk/go/property/equal.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package property
+
+type EqualOption func(*eqOpts)
+
+// See the doc comment for Value.Equals for the effect of EqualRelaxComputed.
+func EqualRelaxComputed(opts *eqOpts) {
+	opts.relaxComputed = true
+}
+
+type eqOpts struct {
+	relaxComputed bool
+}
+
+// Check if two Values are equal.
+//
+// There are two corner cases that need to be called out here:
+//
+//   - Secret equality is enforced. That means that:
+//
+//     {"a", secret: false} == {"a", secret: false}
+//
+//     {"a", secret: true} != {"a", secret: false}
+//
+//     {"b", secret: false} != {"c", secret: false}
+//
+//   - Computed value equality has two different modes. By default, it works like secret
+//     equality, enforcing full structural equality.
+//
+//     If EqualRelaxComputed is true, then computed values are considered equal if they
+//     are of the same type.
```

Comment:
We pretty much never track types of computed values through the system. It would be a damn lot simpler to just treat "computed/unknown" as "we don't know the type". That is actually the case in nearly all cases.